flink-runtime/src/main/java/org/apache/flink/runtime/memory/SharedResources.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.runtime.memory;
+
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.util.function.LongFunctionWithException;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.concurrent.locks.ReentrantLock;
+
+/**
+ * A map that keeps track of acquired shared resources and handles their allocation disposal.
+ */
+final class SharedResources {
+
+	private final ReentrantLock lock = new ReentrantLock();
+
+	private final HashMap<String, LeasedResource<?>> reservedResources = new HashMap<>();
+
+	/**
+	 * Gets the shared memory resource for the given owner and registers a lease. If the resource
+	 * does not yet exist, it will be created via the given initializer function.
+	 *
+	 * <p>The resource must be released when no longer used. That releases the lease. When all leases are
+	 * released, the resource is disposed.
+	 */
+	<T extends AutoCloseable> ResourceAndSize<T> getOrAllocateSharedResource(
+			String type,
+			Object leaseHolder,
+			LongFunctionWithException<T, Exception> initializer,
+			long sizeForInitialization) throws Exception {
+
+		// We could be stuck on this lock for a while, in cases where another initialization is currently
+		// happening and the initialization is expensive.
+		// We lock interruptibly here to allow for faster exit in case of cancellation errors.
+		try {
+			lock.lockInterruptibly();
```

Comment:
Wonder if it could be better when using CAS to ensure only initialized once from previous PR #10416 which refer to the idea of [AtomicSafeInitializer](https://github.com/apache/commons-lang/blob/master/src/main/java/org/apache/commons/lang3/concurrent/AtomicSafeInitializer.java).

Reply:
The lock-free pattern can lead to busy waiting. That is okay if we assume that initialization is always super lightweight (quick).

If the initialization can be more heavyweight in some cases (for example if some systems would allocate / initialize some resources eagerly in that method), the lock is safer in my opinion, as it avoids long busy waiting times. Also, this code path here is not on a hot-per-record code path, so the cost of the lock costs should be okay here.